docker/base/Dockerfile

```diff
@@ -44,19 +46,19 @@ RUN apt-get update && \
     sudo \
     wget \
     zlib1g-dev && \
-  /tmp/deps.sh && \
   cd /tmp/pmacct && \
+  ./ci/deps.sh && \
   export AVRO_LIBS="-L/usr/local/avro/lib -lavro" && \
   export AVRO_CFLAGS="-I/usr/local/avro/include" && \
   ( make maintainer-clean || /bin/true ) && \
-  sh autogen.sh && ./configure --enable-mysql --enable-pgsql     \
+  ./autogen.sh && ./configure --enable-mysql --enable-pgsql     \
                                --enable-sqlite3 --enable-kafka   \
                                --enable-geoipv2 --enable-jansson \
                                --enable-rabbitmq --enable-nflog  \
                                --enable-ndpi --enable-zmq        \
                                --enable-avro --enable-serdes     \
                                --enable-redis --enable-gnutls && \
-  sudo make install && \
+  sudo make -j$(nproc) install && \
```

Comment:
See my comment here https://github.com/pmacct/pmacct/pull/504#discussion_r632854940 on the same topic.

Reply:
I can drop this, but I do feel that in newer CPUs activating parallelism like this makes sense, even more sense given the number of cores are increasing; without this param, unless we are calling gcc with something like `gcc -pipe`, nothing is going to parallelize compilation work, and I don't think that is our case given these tests in different containers (to avoid things being reused):

```bash
root@fbbb4bd280e4:/tmp/pmacct# time make
...
make[1]: Leaving directory '/tmp/pmacct'

real    2m54.137s
user    2m41.248s
sys     0m15.969s
...
root@6af4099cd490:/tmp/pmacct# time make -j 4
...
make[1]: Leaving directory '/tmp/pmacct'
real    1m27.542s
user    3m31.861s
sys     0m22.642s
```

I have been visualizing cpu usage in the two tests, that confirms that without `-j` we just use a single process and cpu core.

Regarding "not being polite" with this strategy, I guess these lines tends to execute in two scenarios:
a) CI system
b) developers/contributors workstations (to test changes, generate local images with local changes...)

For a), I tend to discover and pick all cores available, as CI systems normally isolate different CI jobs and give you dedicated VMs/containers with cgroups, so you can just make the best usage of the resources you see. For b), my very personal view is that I expect processes in my workstation to use as many resources as available to execute as fast or interactive as possible and if, as an user, I want to split my resources to avoid things affecting to each other (better browsing experience while I'm executing heavy work as a compilation process?), it's my job to setup cgroups or other isolation techniques. 

Anyways, if you want a build ARG that defaults to this, but users can override at will, I can work on that. I would not recommend to hardcode the current CI workers specs (they may change, and you are ignoring the specs of users/contributors workstations)

Reply:
>For a), I tend to discover and pick all cores available, as CI systems normally isolate different CI jobs and give you dedicated VMs/containers with cgroups, so you can just make the best usage of the resources you see. 

Using `-j$(nproc)` is a **very bad** idea for _any_ of the options listed.

First you never want to have `make -j $(nproc)` be the number of lcores, even if you are a single user of the machine/vm/whatever. _If_ a `Makefile` has enough translation units (>= nproc) you can severly impact the performance of the machine. Not sure if you ever have done `make -j8` in an 8 lcore machine, but I can tell you I did, and with intensive compilations, and even I/O (mouse, keyboard) of my laptop were laggish, to the point of being unusable. At most you want `nproc-1` to be your workers, assuming this is your only real job.

Second, many (most nowadays) of the times systems are shared; shared servers, VMs, containers. For instance for docker containers, the default is to **NOT limit** the number of CPUs "seen" by the container, unless you explicitely set it:

https://docs.docker.com/config/containers/resource_constraints/#cpu

```
By default, each container’s access to the host machine’s CPU cycles is unlimited.
```

This means that they _see_ the lcores of the host. You can imagine the consequences with more than one container - which is 99% the case in servers hosting docker containers.

See:

```
marc@Mriya:~$ docker run -it debian:buster-slim
root@4faac422c41f:/# nproc 
8
root@4faac422c41f:/# exit
marc@Mriya:~$ docker run -it --cpus=3 debian:buster-slim
root@7becf73d2ed8:/# nproc
8
root@7becf73d2ed8:/# exit
marc@Mriya:~$ docker run -it --cpuset-cpus=1,2 debian:buster-slim
root@f073cb48d1cd:/# nproc
2
root@f073cb48d1cd:/# 
```

The only way to limit `nproc` is to actually use  `--cpuset-cpus`, which is _very_ inconvenient, because that's _the_ list of  lcores (pinning) and not an lcore counter (`--cpus` just computes the equivalent CPU quota, but `nproc` is unaltered). So if you need to "limit" `nproc` you need to start load balancing lcores via `--cpuset-cpus` for all containers, which, again, is _very_ inconvenient. 


If you are still not convinced, I invite you to google a bit on the problems of using `nproc` as an arg to `make -j`. First google result, see the replies:

```
https://github.com/OpenVisualCloud/Dockerfiles/issues/19
```

> Anyways, if you want a build ARG that defaults to this, but users can override at will, I can work on that.

I am advocating for an environament variable _not_ a script argument. If not set define it to a reasonable default (e.g. 2) at the very beginning of the script:

```
if test -z $NUM_WORKERS; then
  NUM_WORKERS=2
fi

#...

#Wherever make is invoked
make -j  $NUM_WORKERS
```

Reply:
Hi Marc. Thanks for the elaborated response:

* Yes, i tend to match `-j` with the number of cores in my system, and I cannot have a VC meeting in parallel, but I don't experience the degradations you are describing (different story is when the `-j` arg exceeds the number of cores in your system, that generates excessive context switching and increases queues). 

* Despite modern servers do have a big number of cpus, I don't think that means we should stick to `1` core, or the fact that increasing this does not come with advantages, which is what I was representing with the previous examples.

* And yes, you are right: I forgot for a second about the complexity of identifying the effective CPU shares you can use inside cgroups, and the fact you see the host cpus, regardless the cpu limits you may be defining. I need to argue, though, that despite most of the CI systems use shared infrastructure, it's strange you land in a non dedicated VM, which gives you the subset of the system you normally are able to consume, but not more. Or, if there are implementations with shared virtual CPUs (I doubt that, given the security implications) and just cgroups for resource management, cpu limits are going to prevent you from doing an unfair usage of the system (a high -j value may mean throttling, to force you to stay under the limits, at the cost of perf degradation compared to a number of parallel jobs that matches the cpu limits for your partition)

* Anyways, regarding the support of a configurable `NUM_WORKERS` param which defaults to 2, I was thinking about a docker-build arg, given the make install lives at the `Dockerfile` level. And we can propagate that to the `ci/deps.sh` script, in the parallel PR

Reply:
> Despite modern servers do have a big number of cpus, I don't think that means we should stick to 1 core, or the fact that increasing this does not come with advantages, which is what I was representing with the previous examples.

Absolutely, and that's why the contrib is useful (thanks again!). To be honest I didn't even check that script, as it was already done for the build CI.

If you want:

* We can set `NUM_WORKERS` by default to 4, which in most workstations and laptops is a sane value nowadays for building dockers locally, and close to the "maximum" real usable `-j` value (I suspect it's 6 or 8 based on my experience in the laptop) 
* We can set the fine tuned  `NUM_WORKERS`  for CI in the Github Actions env.

> cpu limits are going to prevent you from doing an unfair usage of the system (a high -j value may mean throttling, to force you to stay under the limits, at the cost of perf degradation compared to a number of parallel jobs that matches the cpu limits for your partition)

Exactly, which somehow defeats the purpose.

Reply:
>Anyways, regarding the support of a configurable NUM_WORKERS param which defaults to 2, I was thinking about a docker-build arg, given the make install lives at the Dockerfile level. And we can propagate that to the ci/deps.sh script, in the parallel PR

I forgot to answer this one... That'd be sort of the recommended way to do it in `docker`, I agree. If three is an easy way to have a reasonable default when not specified sure, let's go that way. We can invoke it with that build option in the CI (as we want to have it as a Github Actions env variable no matter what)